src/flb_csv.c

```diff
@@ -0,0 +1,316 @@
+        }
+        state->offset++;
+    }
+
+    /* subtract 1 to remove ending double quote */
+    state->length = state->offset - state->start - 1;
+    return 0;
+}
+
+static int parse_csv_field(
+        struct flb_csv_state *state, 
+        const char *data, size_t len)
+{
+    int ret;
+    const char *buf;
+    size_t bufsize;
+    bool buffered = false;
+
+    buf = data;
+    bufsize = len;
+
+    if (state->state == FLB_CSV_STATE_INITIAL) {
+        if (data[state->offset] == '"') {
+            /* advance past opening quote */
+            state->offset++;
+            state->state = FLB_CSV_STATE_STARTED_DQUOTE;
+        } else {
+            state->state = FLB_CSV_STATE_STARTED_SIMPLE;
+        }
+        state->start = state->offset;
+    } else if (state->field_callback) {
+        state->buffered_data = flb_sds_cat(state->buffered_data, data, len);
+        if (!state->buffered_data) {
+            return FLB_CSV_ALLOC_FAILED;
+        }
+        buf = state->buffered_data;
+        bufsize = flb_sds_len(state->buffered_data);
+        buffered = true;
+    }
+
+    switch (state->state) {
+        case FLB_CSV_STATE_STARTED_SIMPLE:
+            ret = parse_simple(state, buf, bufsize);
+            break;
+        case FLB_CSV_STATE_STARTED_DQUOTE:
+        case FLB_CSV_STATE_FOUND_DQUOTE:
+            ret = parse_quoted(state, buf, bufsize);
+            break;
+        default:
+            return FLB_CSV_INVALID_STATE;
+    }
+
+    if (ret) {
+        if (!buffered && ret == FLB_CSV_EOF) {
+            /* not finished, we need to save data in the buffer */
+            state->buffered_data = flb_sds_cat(state->buffered_data, data, len);
+            if (!state->buffered_data) {
+                return FLB_CSV_ALLOC_FAILED;
+            }
+        }
+        return ret;
+    }
+
+    if (state->field_callback) {
+        ret = invoke_field_callback(state, buf, bufsize);
+        if (ret) {
+            return ret;
+        }
+    }
+
+    return ret;
+}
+
+void flb_csv_init(
+        struct flb_csv_state *state,
+        flb_csv_field_parsed_callback field_callback,
+        void *data)
+{
+    state->buffered_data = flb_sds_create("");
+    state->escape_buffer = flb_sds_create("");
+    state->field_callback = field_callback;
+    state->data = data;
+    state->field_count = 0;
+    reset_state(state);
+}
+
+int flb_csv_parse_record(
+        struct flb_csv_state *state,
+        char **bufptr,
+        size_t *buflen,
+        size_t *field_count)
+{
+    char c;
+    int ret;
+    size_t initial_offset;
+    size_t advanced;
+
+    for (;;) {
+        if (!(*buflen)) {
+            return FLB_CSV_EOF;
+        }
+        c = **bufptr;
+        if (state->state == FLB_CSV_STATE_INITIAL) {
+            if (c == '\r') {
+                state->state = FLB_CSV_STATE_FOUND_CR;
+                (*bufptr)++;
+                (*buflen)--;
+                continue;
+            } else if (c == '\n') {
```

Comment:
`else if` must go in the next line: https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#general-requirements